server/src/main/java/org/opensearch/snapshots/SnapshotsService.java

```diff
@@ -1351,69 +1352,75 @@ public void onFailure(Exception e) {
             @Override
             protected void doRun() {
                 final String localNodeId = clusterService.localNode().getId();
-                repository.getRepositoryData(ActionListener.wrap(repositoryData -> {
-                    try {
-                        final IndexMetadata indexMetadata = repository.getSnapshotIndexMetaData(
-                            repositoryData,
+                if (remoteStoreIndexShallowCopy == false) {
+                    executeClone(localNodeId, false);
+                } else {
+                    repository.getRepositoryData(ActionListener.wrap(repositoryData -> {
+                        try {
+                            final IndexMetadata indexMetadata = repository.getSnapshotIndexMetaData(
+                                repositoryData,
+                                sourceSnapshot,
+                                repoShardId.index()
+                            );
+                            final boolean cloneRemoteStoreIndexShardSnapshot = indexMetadata.getSettings()
+                                .getAsBoolean(IndexMetadata.SETTING_REMOTE_STORE_ENABLED, false);
+                            executeClone(localNodeId, cloneRemoteStoreIndexShardSnapshot);
+                        } catch (IOException e) {
+                            logger.warn("Failed to get index-metadata from repository data for index [{}]", repoShardId.index().getName());
+                            failCloneShardAndUpdateClusterState(target, sourceSnapshot, repoShardId);
+                        }
+                    }, this::onFailure));
+                }
+            }
+
+            private void executeClone(String localNodeId, boolean cloneRemoteStoreIndexShardSnapshot) {
+                if (currentlyCloning.add(repoShardId)) {
```

Comment:
I see this matches the original code, but I'm curious what happens to the ActionListener in the case this evaluates false. Is it ever completed?

Reply:
Firstly thanks for the thorough review, @dbwiddis. 

If the `currentlyCloning.add(repoShardId)` returns false, then it means that the underlying shard (with indexId "x" and shard no "y") is being cloned already. Also, runReadyClone method does not return any value to the caller function neither there is any interaction through the action listener. I am not fully sure, but it looks like we do not want to have more than 1 clone operation happening for the same shard. Since we have guard rails that prevent concurrent snapshot operations at cluster level, it probably appears to be defensive coding and not sure if we would ever hit the `currentlyCloning.add(repoShardId)` with value false. Let me know if you think otherwise from code side.

Reply:
OK, looked at it again more closely.  Method naming bit me. `getCloneCompletionListener` actually creates a new listener, so those listeners are never created.  (So nit: should be named `create...()`) 😏 

So this LGTM.